### PR TITLE
Reduce ignored module sanity tests

### PR DESCRIPTION
**Pull request by aminvakil**
##### SUMMARY
Trying to reduce ignores from module sanity tests as possible as I can.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
collection

### CODE REVIEW

File: lib/ansible/modules/apt.py
Comment (Akasurde) [inline lib/ansible/modules/apt.py]:
```suggestion
    aliases: [ update-cache ]
```

To match other parameter styles.

File: lib/ansible/modules/apt.py
Comment (Akasurde) [inline lib/ansible/modules/apt.py]:
```suggestion
    aliases: [ default-release ]
```

File: lib/ansible/modules/apt.py
Comment (Akasurde) [inline lib/ansible/modules/apt.py]:
```suggestion
    aliases: [ allow-unauthenticated ]
```

File: lib/ansible/modules/apt_key.py
Comment (Akasurde) [inline lib/ansible/modules/apt_key.py]:
```suggestion
        aliases: [ filename ]
```

File: lib/ansible/modules/apt_repository.py
Comment (Akasurde) [inline lib/ansible/modules/apt_repository.py]:
```suggestion
        aliases: [ update-cache ]
```

File: lib/ansible/modules/apt_repository.py
Comment (Akasurde) [inline lib/ansible/modules/apt_repository.py]:
```suggestion
            - If python-apt Python library is not installed this module will not work.
```

File: lib/ansible/modules/package_facts.py
Comment (Akasurde) [inline lib/ansible/modules/package_facts.py:413]:
```suggestion
    module = AnsibleModule(argument_spec=dict(manager={'type': 'list', 'elements': 'str', 'default': ['auto'], choices=['apt', 'auto', 'pacman', 'pkg', 'portage', 'rpm']},
```

**Review by Akasurde — CHANGES_REQUESTED** [review]
@aminvakil Thanks for the contribution. Could you please rebase and apply the suggestions? @Andersson007 @felixfontein Could you please review as well? Thanks in advance.

**Comment by Akasurde** [comment]
@aminvakil Could you please rebase using https://docs.ansible.com/ansible/latest/dev_guide/developing_rebasing.html#rebasing-a-pull-request?

File: lib/ansible/modules/package_facts.py
Comment (aminvakil) [reply lib/ansible/modules/package_facts.py:413]: I'm not sure If I've changed correctly as your suggestion, could you please review this?
https://app.shippable.com/github/ansible/ansible/runs/169575/2/tests

Comment (aminvakil) [reply lib/ansible/modules/package_facts.py:413]: Also this:
https://app.shippable.com/github/ansible/ansible/runs/169575/1/tests

Comment (Akasurde) [reply lib/ansible/modules/package_facts.py:413]:
```suggestion
    module = AnsibleModule(argument_spec=dict(manager={'type': 'list', 'elements': 'str', 'default': ['auto'], choices: ['apt', 'auto', 'pacman', 'pkg', 'portage', 'rpm']},
```

Comment (Akasurde) [reply lib/ansible/modules/package_facts.py:413]: Oops. My bad. Changed it.

File: lib/ansible/modules/package_facts.py
Comment (Akasurde) [inline lib/ansible/modules/package_facts.py]:
```suggestion
    module = AnsibleModule(argument_spec=dict(manager={'type': 'list', 'elements': 'str', 'default': ['auto'], choices: ['apt', 'auto', 'pacman', 'pkg', 'portage', 'rpm']}, 
                                              strategy={'choices': ['first', 'all'], 'default': 'first'}),
```

No need to remove `strategy`

Comment (aminvakil) [reply lib/ansible/modules/package_facts.py]: Oops. My bad. Changed it.

Comment (aminvakil) [reply lib/ansible/modules/package_facts.py]: Shouldn't choices be in '' ?
It got fixed, if anything else needs to be changed, please let me know.

**Review by Akasurde — APPROVED** [review]
LGTM, I will wait for others to comment.

File: lib/ansible/modules/apt_key.py
Comment (felixfontein) [inline lib/ansible/modules/apt_key.py]: Why did you add this new alias (both in docs and in code)?

File: lib/ansible/modules/apt_repository.py
Comment (felixfontein) [inline lib/ansible/modules/apt_repository.py]: This changes behavior. If `mode` is `None`, in https://github.com/ansible/ansible/blob/bad591cfbe3e254f02768fa4af7679a88af2343e/lib/ansible/modules/apt_repository.py#L334 `this_mode` will be `None` (since `module.params['mode'] == None`) and thus `self.module.set_mode_if_different(filename, this_mode, False)` in the next line will be a no-op.

File: lib/ansible/modules/subversion.py
Comment (felixfontein) [inline lib/ansible/modules/subversion.py]: `dest` is only required for some values of `state`. It should not be marked as required.

File: lib/ansible/modules/yum_repository.py
Comment (felixfontein) [inline lib/ansible/modules/yum_repository.py]: The defaults here also change behavior IMO. Even if these are the default values used by yum, this could lead to repository files being changed.

File: lib/ansible/modules/apt_key.py
Comment (aminvakil) [reply lib/ansible/modules/apt_key.py]: Few lines below the line you've mentioned there is:
```
filename = module.params['file']
```
And this was in `test/sanity/ignore.txt`:
```
lib/ansible/modules/apt_key.py validate-modules:parameter-type-not-in-doc
lib/ansible/modules/apt_key.py validate-modules:undocumented-parameter
```
To remove these two lines in `ignore.txt` I added this alias (both in docs and in code).

File: lib/ansible/modules/apt_repository.py
Comment (aminvakil) [reply lib/ansible/modules/apt_repository.py]: So I assume `default: '0644'` has to be removed from docs instead of adding it in module argument_spec?

File: lib/ansible/modules/subversion.py
Comment (aminvakil) [reply lib/ansible/modules/subversion.py]: I couldn't think of a situation where this module would work without specifying `dest`. Also `state` is not defined in this module, so I'm not sure what variable you meant.
Anyway I'm going to remove `required=True` for `dest` if you think it's not required and ignore it in `ignore.txt`.

File: lib/ansible/modules/yum_repository.py
Comment (aminvakil) [reply lib/ansible/modules/yum_repository.py]: So maybe this has to remove from docs too instead of adding it to argument_spec?

File: lib/ansible/modules/apt_repository.py
Comment (Andersson007) [reply lib/ansible/modules/apt_repository.py]: @aminvakil yes, imo, a general rule should be like if there are deviations between doc and arg_spec, changing the doc would be preferable because it doesn't affect actual module's behavior, so safer

**Comment by Andersson007** [comment]
@aminvakil thanks for doing this! Fixing such general stuff is highly appreciated!

File: lib/ansible/modules/apt_key.py
Comment (felixfontein) [reply lib/ansible/modules/apt_key.py]: `filename` is a Python variable inside the module. It is not visible in any way to the user. The module option is called `file`.

I don't know what exactly the warnings were about, but it was not that a Python variable was not mentioned in argspec.

File: lib/ansible/modules/apt_repository.py
Comment (felixfontein) [reply lib/ansible/modules/apt_repository.py]: If it would be changed the other way, it would need a `breaking_change` changelog and should better be done by the maintainer(s), since they (hopefully) know what was intended. Maybe in this case we still need a changelog fragment (`bugfixes`?) because some people might have relied on the `0644` assumption and didn't notice it didn't happen.

File: lib/ansible/modules/subversion.py
Comment (felixfontein) [reply lib/ansible/modules/subversion.py]: See https://github.com/ansible/ansible/blob/bad591cfbe3e254f02768fa4af7679a88af2343e/lib/ansible/modules/subversion.py#L284-L285

The `required` should be removed from documentation (resp. needs to be mentioned in the text there instead, in which case exactly the option is required).

File: lib/ansible/modules/yum_repository.py
Comment (felixfontein) [reply lib/ansible/modules/yum_repository.py]: Yes. This should probably be mentioned in the description, that yum's default value is `true`, but not in the `default:` value.

File: lib/ansible/modules/apt_repository.py
Comment (Andersson007) [inline lib/ansible/modules/apt_repository.py]: what about the comment in arg_spec ?
```
            # This should not be needed, but exists as a failsafe
            install_python_apt=dict(type='bool', default=True),
```
i think this parameter was removed from the doc intentionally

File: lib/ansible/modules/dpkg_selections.py
Comment (Andersson007) [inline lib/ansible/modules/dpkg_selections.py]:
```suggestion
            - Name of the package.
```
while you're here:)

File: lib/ansible/modules/apt_repository.py
Comment (felixfontein) [reply lib/ansible/modules/apt_repository.py]: It's a good question. This comment has been there like forever (I checked quite some years back in history). For this we probably need core team input, or input from someone who knows the history of this module well.

**Comment by aminvakil** [comment]
> @aminvakil thanks for doing this! Fixing such general stuff is highly appreciated!

You're welcome, it's a great pleasure working on ansible.

File: lib/ansible/modules/apt_repository.py
Comment (aminvakil) [reply lib/ansible/modules/apt_repository.py]: On minimal Debian `python-apt` is not installed for example and right now ansible shows a warning **after** it installs `python-apt` on the server being managed by ansible which would be a problem for someones which are sensitive about packages installed on their server.

I mean there would be a situation where a sysadmin would prefer installing a package herself instead of installing `python-apt` and the package from the playbook.

It should be mentioned in documentation that by default this module installs `python-apt` and people should have the choice whether to disable installing it or not.

File: lib/ansible/modules/apt_key.py
Comment (aminvakil) [reply lib/ansible/modules/apt_key.py]: This was added because of this in `ignore.txt`:
`lib/ansible/modules/apt_key.py validate-modules:mutually_exclusive-unknown`
In order to remove this ignored validation I added `filename` as an alias to `file`.
Right now it has not been added to `file` as an alias as ignore line is present in ignore.txt.
Could you please let me know if it needs to change?

File: lib/ansible/modules/apt_repository.py
Comment (aminvakil) [reply lib/ansible/modules/apt_repository.py]: Is this OK?
https://github.com/ansible/ansible/pull/70319/commits/2b346b58ff8b4aa50619220bb3e2515069acd979

File: lib/ansible/modules/subversion.py
Comment (aminvakil) [reply lib/ansible/modules/subversion.py]: Is this OK?
https://github.com/ansible/ansible/pull/70319/commits/92c57aff202adf76493fab6e38e5e6966dada507

File: lib/ansible/modules/dpkg_selections.py
Comment (aminvakil) [reply lib/ansible/modules/dpkg_selections.py]: Fixed.

File: lib/ansible/modules/yum_repository.py
Comment (aminvakil) [reply lib/ansible/modules/yum_repository.py]: Could you please check the description format and see if it's OK or not?

File: lib/ansible/modules/apt_key.py
Comment (felixfontein) [reply lib/ansible/modules/apt_key.py]: The error is that `mutually_exclusive` contains `filename`, which is not an option. It should contain `file` instead. (Adding an alias `filename` would be wrong, since then `mutually_exclusive` would not work correctly.)

This is a potentially breaking change, so it needs a changelog fragment and a porting guide entry.

File: lib/ansible/modules/apt_repository.py
Comment (felixfontein) [reply lib/ansible/modules/apt_repository.py]: I think so.

File: lib/ansible/modules/subversion.py
Comment (felixfontein) [inline lib/ansible/modules/subversion.py]:
```suggestion
      - The destination directory must be specified unless I(checkout=no), I(update=no), and I(export=no).
```

File: lib/ansible/modules/subversion.py
Comment (felixfontein) [reply lib/ansible/modules/subversion.py]: Yes, almost (see my below comment).

File: lib/ansible/modules/yum_repository.py
Comment (felixfontein) [reply lib/ansible/modules/yum_repository.py]: I'd use `C(true)` for the value; besides that, yes.

File: lib/ansible/modules/apt_key.py
Comment (aminvakil) [reply lib/ansible/modules/apt_key.py]: OK, so is it appropriate if this gets fixed in this PR or leave it be this way for now?

File: lib/ansible/modules/apt_repository.py
Comment (felixfontein) [reply lib/ansible/modules/apt_repository.py]: I know what you mean, and I think it's better to document this properly, but I think this PR is the wrong place for it since this needs further discussion, which would delay the whole PR unnecessarily. It's better to create a new PR for only this change (once this one is merged f.ex., or already now if you don't mind that you might have to resolve conflicts), so that it can be properly discussed.

File: lib/ansible/modules/yum_repository.py
Comment (aminvakil) [reply lib/ansible/modules/yum_repository.py]: Could you please explain more?

File: lib/ansible/modules/apt_repository.py
Comment (aminvakil) [reply lib/ansible/modules/apt_repository.py]: Fine, I will revert the changes to way `devel` branch is now and will create a new PR and put the link here.

File: lib/ansible/modules/subversion.py
Comment (aminvakil) [reply lib/ansible/modules/subversion.py]: Fixed in https://github.com/ansible/ansible/pull/70319/commits/65503c2b7a98843dbf867b8a9fd90d2f7b43c6e1.

File: lib/ansible/modules/apt_repository.py
Comment (aminvakil) [reply lib/ansible/modules/apt_repository.py]: As @felixfontein said it got removed from this PR and it can be further discussed in https://github.com/ansible/ansible/pull/70462.

File: lib/ansible/modules/yum_repository.py
Comment (aminvakil) [reply lib/ansible/modules/yum_repository.py]: > I'd use `C(true)` for the value; besides that, yes.

Fixed in https://github.com/ansible/ansible/pull/70319/commits/66ec490a8acf5441329564f6c6343c5b14c68f50.

**Comment by aminvakil** [comment]
ready_for_review

File: lib/ansible/modules/apt_key.py
Comment (felixfontein) [reply lib/ansible/modules/apt_key.py]: Both is fine, but if you address it, you definitely need a changelog fragment for that change. (Since it potentially breaks people's roles/playbooks if they implicitly relied on that specifying `file` together with other options works. It's probably better to have this in a separate PR, that cannot be backported to stable-2.10.)

File: changelogs/fragments/70319-reduce-ignored-module-sanity-tests.yml
Comment (felixfontein) [inline changelogs/fragments/70319-reduce-ignored-module-sanity-tests.yml]:
```suggestion
  - apt_repository - fixes ``mode`` doc to remove ineffective default (https://github.com/ansible/ansible/pull/70319).
```
I'm not sure whether there should be a changelog fragment for this change at all, though, since this doesn't change detectable behavior for users. That's just my personal opinion though :)

**Review by felixfontein — COMMENTED** [review]
Besides that, LGTM.

File: lib/ansible/modules/apt_key.py
Comment (aminvakil) [reply lib/ansible/modules/apt_key.py]: Will fix it in another PR then.

Comment (aminvakil) [reply lib/ansible/modules/apt_key.py]: Could be further discussed in https://github.com/ansible/ansible/pull/70492.

**Review by felixfontein — APPROVED** [review]

**Review by Andersson007 — APPROVED** [review]
LGTM

**Comment by aminvakil** [comment]
@Akasurde @Andersson007 @felixfontein Thanks for reviewing this.
Special thanks fo @felixfontein for all the work on my recent pull requests, you have taught me a lot during these past few weeks, on this pull request and the other pull requests from https://github.com/ansible-collections/community.general.

**Comment by felixfontein** [comment]
@Akasurde can you take another look?

**Comment by Akasurde** [comment]
@aminvakil Could you please rebase this branch?

**Comment by aminvakil** [comment]
@Akasurde Made some mistakes in rebasing, I'm fixing the problems now.

**Comment by aminvakil** [comment]
@Akasurde Could you please review now?

File: lib/ansible/modules/apt_key.py
Comment (felixfontein) [inline lib/ansible/modules/apt_key.py:270]: Options should not be removed without a deprecation cycle.

**Review by felixfontein — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/apt_key.py
Comment (aminvakil) [reply lib/ansible/modules/apt_key.py:270]: @felixfontein Could you please check https://github.com/ansible/ansible/pull/70319/commits/cfcacaa1cf4d98fad68120b270b545bf0148fbb4 and these minor fixes in https://github.com/ansible/ansible/pull/70319/commits/c7c3a51e81761df7e94db0dc307f74627e0d2851 and https://github.com/ansible/ansible/pull/70319/commits/ea20be4462c4bd8d47082fa433e425f115990d8a?

File: lib/ansible/modules/apt_key.py
Comment (felixfontein) [inline lib/ansible/modules/apt_key.py]:
```suggestion
            key=dict(type='str', removed_in_version='2.14', removed_from_collection='ansible.builtin'),
```

File: lib/ansible/modules/apt_key.py
Comment (felixfontein) [inline lib/ansible/modules/apt_key.py]: Remove this `if`.

File: changelogs/fragments/70319-reduce-ignored-module-sanity-tests.yml
Comment (felixfontein) [inline changelogs/fragments/70319-reduce-ignored-module-sanity-tests.yml]:
```suggestion
  - apt_key - the paramater ``key`` does not have any effect, has been deprecated and will be removed in ansible-base version 2.14 (https://github.com/ansible/ansible/pull/70319).
```
The deprecation cycle is four versions (changes in devel will end up in 2.11, so this will get removed in 2.14).

**Comment by aminvakil** [comment]
I can't understand why I get this error:
https://app.shippable.com/github/ansible/ansible/runs/170458/5/tests

File: changelogs/fragments/70319-reduce-ignored-module-sanity-tests.yml
Comment (felixfontein) [inline changelogs/fragments/70319-reduce-ignored-module-sanity-tests.yml]:
```suggestion
  - apt_key - the paramater ``key`` does not have any effect, has been deprecated and will be removed in ansible-base version 2.14 (https://github.com/ansible/ansible/pull/70319).
```

**Comment by felixfontein** [comment]
@aminvakil the error comes because in `devel` branch, `apk` was added to the list in DOCUMENTATION (https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/package_facts.py#L24). Since shippable automatically rebases against `devel`, and you didn't touch that list, the resulting file processed by ansible-test has a longer list in DOCUMENTATION (the one from `devel`) than in the argument spec (the list you added).

**Comment by felixfontein** [comment]
(And because you didn't touch the documentation list, there's no conflict.)

**Comment by aminvakil** [comment]
> @aminvakil the error comes because in `devel` branch, `apk` was added to the list in DOCUMENTATION (https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/package_facts.py#L24). Since shippable automatically rebases against `devel`, and you didn't touch that list, the resulting file processed by ansible-test has a longer list in DOCUMENTATION (the one from `devel`) than in the argument spec (the list you added).

@felixfontein Thank you for explanation, so aside from that I think other issues that you've mentioned has been resolved.

**Comment by aminvakil** [comment]
ready_for_review

File: lib/ansible/modules/package_facts.py
Comment (felixfontein) [inline lib/ansible/modules/package_facts.py]: This is the only change I'm not really sure whether this behavior is wanted. Right now, even though the documentation forces you to stick to this list of values, you can specify something else. The code below valides this option, and potentially allows new values that are not yet listed in `choices`, as long as support for them has been implemented.

Also, right now, `PACMAN` for example is also accepted, and converted to `pacman` by the module (https://github.com/ansible/ansible/blob/c6ce08bb5d93f7d625905d031a4567d54f00a961/lib/ansible/modules/package_facts.py#L419). With this change, this is no longer possible.

Comment (aminvakil) [reply lib/ansible/modules/package_facts.py]: It seems that ignoring `doc-choices-do-not-match-spec` is reasonable because of the module implementation, I'm going to ignore it in this PR.

Comment (felixfontein) [reply lib/ansible/modules/package_facts.py]: You still need to remove the `choices` from `argument_spec`.

**Review by felixfontein — APPROVED** [review]
LGTM

**Comment by mkrizek** [comment]
!waffling core_review

**Comment by felixfontein** [comment]
@Akasurde can you take another look? I think this is ready to merge.

**Comment by mkrizek** [comment]
!waffling needs_revision

**Comment by aminvakil** [comment]
bot_status

**Comment by aminvakil** [comment]
ready_for_review

File: lib/ansible/modules/apt_repository.py
Comment (bmillemathias) [inline lib/ansible/modules/apt_repository.py:38]: better to remove approximate comment ' (probably 0644)'.
Also putting a value that likely to change over time can be wrong, except if we could extract that from a constant.

**Review by bmillemathias — COMMENTED** [review]
Good amount of work

File: lib/ansible/modules/apt_repository.py
Comment (aminvakil) [reply lib/ansible/modules/apt_repository.py:38]: This has been discussed on https://github.com/ansible/ansible/pull/70319#discussion_r448980070 too.

But now that you mentioned this, I haven't seen this line before:
https://github.com/ansible/ansible/blob/db519bc17c1ff0c6e095d022bfe988eb87630090/lib/ansible/modules/apt_repository.py#L143
https://github.com/ansible/ansible/blob/db519bc17c1ff0c6e095d022bfe988eb87630090/lib/ansible/modules/apt_repository.py#L322

It seems that default `mode` is `0644` and maybe that was why it has been mentioned on `mode` doc, am I right?

If so I think adding `default: '0644'` on doc would be appropriate because it explains the module behaviour as it is.

Comment (aminvakil) [reply lib/ansible/modules/apt_repository.py:38]: Also these:
https://github.com/ansible/ansible/commit/127b854d2c0eda8de6954cc5eac46e6681577c65
https://github.com/ansible/ansible/commit/a5b80464df92c2ebdc7ca73159763d10d4ed7068#diff-355558b0484f1dfbf4d532cbc563c9bdR136

File: lib/ansible/modules/apt_repository.py
Comment (felixfontein) [inline lib/ansible/modules/apt_repository.py]: Please undo this change.

**Review by felixfontein — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/apt_repository.py
Comment (felixfontein) [reply lib/ansible/modules/apt_repository.py:38]: `this_mode = self.module.params.get('mode', DEFAULT_SOURCES_PERM)` returns `None` if `mode` is not provided as a module parameter. Changing the `default` of this option is thus a breaking change.

Comment (aminvakil) [reply lib/ansible/modules/apt_repository.py:38]: @felixfontein Thanks, I didn't know this. What's your opinion on removing `(probably 0644)`?
It can be helpful for those who care about their `apt_repository` file permission, also it gives a hint that they can use setuid, setgid and sticky bits and they should put a leading zero if they don't want them, so maybe changing it to `(for example 0644)` would be better than removing it.

Comment (felixfontein) [reply lib/ansible/modules/apt_repository.py:38]: I think it is ok to leave it for now. I guess the core team needs to decide whether they want to fix the default to be `0644` (that would require a behavior change), or remove it completely.

**Review by felixfontein — APPROVED** [review]

File: lib/ansible/modules/yum_repository.py
Comment (s-hertel) [inline lib/ansible/modules/yum_repository.py:647]: Is this a change in behavior to match the docs?

Comment (samdoran) [reply lib/ansible/modules/yum_repository.py:647]: Looks like that is the case.

**Comment by aminvakil** [comment]
@felixfontein @Andersson007 @Akasurde @bmillemathias @s-hertel @samdoran Thanks for reviewing!

**Pull request merged**
